llvm/lib/Target/AIE/AIEPostSelectOptimize.cpp

```diff
@@ -438,6 +440,140 @@ bool duplicateAdressingRegs(MachineBasicBlock &MBB, MachineRegisterInfo &MRI) {
   return tryToDuplicateLoadUse(LoadUses, NonLoadUses, MRI, TII);
 }
 
+using Operation = AIEBaseInstrInfo::AbstractOp::OperationType;
+using AbstractOp = AIEBaseInstrInfo::AbstractOp;
+using OptionalOp = std::optional<AbstractOp>;
+
+bool getGlobalValue(const MachineInstr *MI,
+                    SmallPtrSet<const Value *, 4> &GVSet,
+                    MachineRegisterInfo &MRI) {
+
+  MI = getDefIgnoringCopiesAndBitcasts(MI->getOperand(1).getReg(), MRI);
+
+  // We need an instruction that explicitly moves a global
+  // to a register (move immediate).
+  if (MI->getNumOperands() != 2)
+    return false;
+
+  const MachineOperand *MO = MI->uses().begin();
+
+  if (MO->isGlobal()) {
+    GVSet.insert(MO->getGlobal());
+    return true;
+  }
+
+  return false;
+}
+
+// Recognize BROADCAST (GLOBAL)
+bool visitVBroadcast(const AbstractOp &VBroadcast, const AIEBaseInstrInfo *TII,
+                     MachineRegisterInfo &MRI,
+                     SmallPtrSet<const Value *, 4> &GVSet) {
+  assert(VBroadcast.Type == Operation::VECTOR_BROADCAST &&
+         "Wrong abstract operation.");
+  return getGlobalValue(MRI.getVRegDef(VBroadcast.ScalarSrcRegs[0]), GVSet,
+                        MRI);
+}
+
+// Recognize SELECT ((SELECT | BROADCAST), ((SELECT | BROADCAST)))
+bool visitVSelect(const AbstractOp &VSelect, const AIEBaseInstrInfo *TII,
+                  MachineRegisterInfo &MRI,
+                  SmallPtrSet<const Value *, 4> &GVSet) {
+
+  assert(VSelect.Type == Operation::VECTOR_SELECT &&
+         "Wrong abstract operation.");
+
+  bool Success = true;
+  for (const Register Reg : VSelect.VectorSrcRegs) {
+    OptionalOp VOp = TII->parseAbstractOp(*MRI.getVRegDef(Reg));
+    // We don't recognize this! Here we cannot have anything different
+    // from SELECT or BROADCAST.
+    if (!VOp)
+      return false;
+    if (VOp->Type == Operation::VECTOR_SELECT)
+      Success &= visitVSelect(*VOp, TII, MRI, GVSet);
+    else if (VOp->Type == Operation::VECTOR_BROADCAST)
+      Success &= visitVBroadcast(*VOp, TII, MRI, GVSet);
+    else
+      return false;
+  }
+
+  return Success;
+}
+
+// Recognize ADD ((ADD | SELECT), Unknown)
+bool visitAddressVector(const AbstractOp &VOp, const AIEBaseInstrInfo *TII,
+                        MachineRegisterInfo &MRI,
+                        SmallPtrSet<const Value *, 4> &GVSet) {
+
+  if (VOp.Type == Operation::VECTOR_SELECT)
+    return visitVSelect(VOp, TII, MRI, GVSet);
+  if (VOp.Type != Operation::VECTOR_ADD)
+    return false;
+
+  return any_of(VOp.VectorSrcRegs, [&](Register SrcReg) {
+    OptionalOp VOp = TII->parseAbstractOp(*MRI.getVRegDef(SrcReg));
+    return !VOp || visitAddressVector(*VOp, TII, MRI, GVSet);
```

Comment:
I think I made a mistake when suggesting that cod. That should probably be `VOp && visitAddressVector(*VOp, TII, MRI, GVSet)`. I.e. we are fine as long as one of the source registers could be successfully traced to an address vector. What do you think?

Reply:
If one can be traced back, we are fine! I will update.